src/parser/core/Module.ts

```diff
@@ -0,0 +1,273 @@
+import {cloneDeep} from 'lodash'
+import 'reflect-metadata'
+
+import {Event, Ability} from 'fflogs'
+import Parser from './Parser'
+
+export const DISPLAY_ORDER = {
+	TOP: 0,
+	DEFAULT: 50,
+	BOTTOM: 100,
+}
+
+export function dependency(target: Module, prop: string) {
+	const dependency = Reflect.getMetadata('design:type', target, prop)
+	const constructor = target.constructor as typeof Module
+
+	// Make sure we're not modifying every single module
+	if (constructor.dependencies === Module.dependencies) {
+		constructor.dependencies = []
+	}
+
+	// If the dep is Object, it's _probably_ from a JS file. Fall back to simple handling
+	if (dependency === Object) {
+		constructor.dependencies.push(prop)
+		return
+	}
+
+	// Check that the dep is actually a module
+	if (!Module.isPrototypeOf(dependency)) {
+		throw new Error(`${constructor.name}'s dependency \`${prop}\` is invalid. Expected \`Module\`, got \`${dependency.name}\`.`)
+	}
+
+	constructor.dependencies.push({
+		handle: dependency.handle,
+		prop,
+	})
+}
+
+interface MappedDependency {
+	handle: string
+	prop: string
+}
+
+type Filter<T extends Event> = Partial<T> & Partial<{
+	abilityId: Ability['guid']
+	to: 'player' | 'pet' | T['targetID']
+	by: 'player' | 'pet' | T['sourceID']
+}>
+
+type HookCallback<T extends Event> = (event: T) => void
+
+interface Hook<T extends Event> {
+	events: T['type'][]
+	filter: Filter<T>
+	callback: HookCallback<T>
+}
+
+export default class Module {
+	static dependencies: Array<string | MappedDependency> = []
```

Comment:
I prefer `(string | MappedDependency)[]`, but whatever goes here.

I also prefer using `ReadonlyArray` / `ReadonlyMap` / `ReadonlySet` wherever possible, but the decorator has to be able to `.push` to it so this one can't be readonly.